tests/test_lion8b.py

```diff
@@ -0,0 +1,549 @@
+        params.append(p)
+
+    # create optimizer and have it step so that state gets populated
+    opt = Lion8bit(params,
+                   compress_state_dict=quantized_state,
+                   error_correction=use_errors)
+    if device == 'cpu':
+        with pytest.raises(NotImplementedError):
+            opt.step()
+        return
+    else:
+        opt.step()
+    opt.zero_grad()
+
+    # copy state dict into a new instance
+    state_dict = opt.state_dict()
+    opt_new = Lion8bit(params,
+                       compress_state_dict=quantized_state,
+                       error_correction=use_errors)
+    opt_new.load_state_dict(state_dict)
+
+    for p in params:
+        d_orig = opt.state[p]
+        d_new = opt_new.state[p]
+        assert list(d_orig.keys()) == list(d_new.keys())
+        mom_orig = d_orig['exp_avg']
+        mom_new = d_new['exp_avg']
+        if quantized_state:
+            # Optimizer load_state_dict insists on converting scales to
+            # dtype of param, which is lossy for bf16 params.
+            # Ideally we'd require == for everything but it's less complexity
+            # to just relax the bf16 test
+            assert torch.all(mom_orig.quantized == mom_new.quantized)
+            if dtype == torch.bfloat16:
+                torch.testing.assert_close(mom_orig.scales,
+                                           mom_new.scales,
+                                           atol=1e-3,
+                                           rtol=1e-2)
+            else:
+                assert torch.all(mom_orig.scales == mom_new.scales)
+
+        torch.testing.assert_close(mom_orig.materialize(),
+                                   mom_new.materialize(),
+                                   atol=1. / (2 * 127),
+                                   rtol=np.inf)
+        if use_errors and (dtype != torch.float32):
+            torch.testing.assert_close(d_orig['errors'], d_new['errors'])
+
+
+class _DummyModule(nn.Module):
+
+    def __init__(self, device: str, dtype: torch.dtype):
+        super().__init__()
+        self.linear0 = nn.Linear(4, 3, device=device, dtype=dtype)
+        self.linear1 = nn.Linear(3, 4, device=device, dtype=dtype)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:  # type:ignore
+        return self.linear1(self.linear0(x))
+
+
+_FULL_STATE = fsdp.StateDictType.FULL_STATE_DICT
+_SHARDED_STATE = fsdp.StateDictType.SHARDED_STATE_DICT
+_LOCAL_STATE = fsdp.StateDictType.LOCAL_STATE_DICT
+
+
+# run just this test with:
+# python3 -m composer.cli.launcher -n 2 --master_port 26000 -m pytest -m gpu tests/test_lion8b.py::test_fsdp_save_load  # noqa
+@pytest.mark.gpu
+@pytest.mark.world_size(2)
+@pytest.mark.parametrize('dtype', _FLOAT_DTYPES)
+@pytest.mark.parametrize('use_errors', [False, True])
+@pytest.mark.parametrize('state_sharding',
+                         [_FULL_STATE, _SHARDED_STATE, _LOCAL_STATE])
+def test_fsdp_save_load(dtype: torch.dtype, use_errors: bool,
```

Comment:
nice. really good coverage test! any chance you can add a test similar to this where you actually pass mod and opt to `Trainer` and train for a batch or two?

Reply:
How important do you estimate this is? We've used it for a bunch of training runs and, given the way the optimizer works, FSDP should really have no idea it's different from any other optimizer. Trying to scope this down because it's already a 1kloc PR that's been "done" for a week already and research people are using it from branches and stuff right now

Reply:
ok if you've done manual tests then it's probably fine